### DISCUSSION TITLE
Speed loss when adding categories

### DISCUSSION THREAD

**Discussion by jagrundy** [scikit-hep/pyhf#1677] [Q&A]
Hi!

In our ATLAS analysis, we have noticed a significance speed difference between our pyhf limit setting framework and another independent framework, and wanted your thoughts/advice. I'm a member of the HH → 4b analysis and one of results is a limit on HH production cross-section at various κλ (the Higgs self-coupling modifier) points. We have a number of categories in the analysis, the addition of which has slowed our pyhf-based [limit setting framework](https://gitlab.cern.ch/hh4b/hh4b-vbf-limits) down significantly. Put simply, we use a root finder and `pyhf.infer.hypotest` to find the μ corresponding to CLs = 0.05. For example, for  κλ = 1, the time taken to get an expected limit on μ with error bands is ~25 hours 39 minutes. An alternative framework used by the HH combination team, [quickstats](https://pypi.org/project/quickstats/), which uses ROOT, extracted the same limit in ~79 seconds. We found this difference very surprising! Please let us know your thoughts and if you want any more information.

Cheers!
James

**Comment by alexander-held** [comment]
Hi James, there are a number of things that may have an impact here. As a first step, I would suggest comparing the runtime of a single maximum-likelihood estimate (MLE) fit. A few more thoughts:
-  `pyhf` does not prune gammas (modifiers for MC statistical uncertainty): https://github.com/scikit-hep/pyhf/issues/662. If you have many bins, and apply pruning, that may significantly reduce the number of parameters in the ROOT setup.
- I don't know the `quickstats` limit-finding implementation, but there are some scripts used in ATLAS that are extremely efficient at ROOT finding by using features of the problem. How may fits do you need for convergence in the various setups? What threshold are you using to determine convergence? Usually the POI resolution does not need to be very fine, allowing to potentially stop early.
- Have you tried different `pyhf` backends? Which minimizer are you using? That may significantly change the runtime.

When you say "categories", do you mean channels in `pyhf` (= phase space regions)? That may point to the gamma point above.

If you are interested in another datapoint, you may also want to check out the implementation in `cabinetry`: [code](https://github.com/scikit-hep/cabinetry/blob/b89bb09b97608e984864cde6d45f9b8901612a06/src/cabinetry/fit/__init__.py#L539-L641) / [docs](https://cabinetry.readthedocs.io/en/latest/api.html#cabinetry.fit.limit). This uses a few things to speed up the procedure, but is typically still surpassed by optimized implementations (some more tricks need to be ported to this still).

**Reply by jagrundy** [reply]
Hi Alex, just to answer a few of these questions - the `quickstats` framework also doesn't prune, so it's likely this isn't the problem. Regarding `quickstats`, I believe the tolerance for the scan is 0.001 rather than 1e-6, which is used by pyhf's SciPy optimizer. I'm looking to try this lower tolerance in pyhf now. On the backends and minimizers, we've tried varying this but without any real change in performance.

**Reply by alexander-held** [reply]
Regarding the pruning: What are you setting for the `RelErrorThreshold` in the stat error configuration in the HistFactory xmls? Unless this is set to an extremely low value (lower than the actual stat error in any bin), the ROOT implementation will automatically drop gammas for bins below threshold, while `pyhf` does not (yet) follow this behavior.

For the tolerance: I did not mean the tolerance for the estimated distance to the minimum (to stop a given fit early), but the tolerance in the POI value when finding the CLs = 0.05 crossing. That said, also the tolerance you are referring to may make quite a bit of a difference in the amount of steps required during minimization.

**Reply by jagrundy** [reply]
I haven't used HF before so am unsure but the .xml file I use with `quickstats` doesn't have `RelErrorThreshold` listed in.

The latter tolerance doesn't seem to problem. The framework hangs for a long period at each step in the POI root finder for CLs = 0.05. It seems to be getting stuck in the optimizer for e.g. `pyhf.infer.test_statistics.qmu_tilde`.

**Reply by alexander-held** [reply]
This setting should appear in the per-region xml files (not the top level file). It can look like this:
```xml
<StatErrorConfig RelErrorThreshold="0.01" ConstraintType="Poisson" />
```
The default for this is 5%, which is rather high. Unless you don't use MC statistical uncertainties at all (e.g.
```xml
<StatError Activate="True"  InputFile=""  HistoName=""  HistoPath=""  />
```
in the xmls, or `staterror` type modifiers in the JSON workspace), this could make a difference between ROOT and pyhf.

**Reply by jagrundy** [reply]
Okay, in the per-region xml files I've found e.g.
```
<StatError Activate="True" HistoName="histggf_16_4b_lowdEta_Xhh_0_0.95_ggfSig_ggf_16_4b_lowdEta_Xhh_0_0.95_ggfSig_stat_ggf_16_4b_lowdEta_Xhh_0_0.95" />
```

**Reply by jagrundy** [reply]
However, we do use lots of stat errors in our JSON.

**Reply by alexander-held** [reply]
If you have no `StatErrorConfig`, then in ROOT you would automatically prune all gammas below 5%. That could potentially affect most of your bins, and then cause your `pyhf` fit to have ~500 more parameters than the fit with ROOT.

**Reply by jagrundy** [reply]
When I search for `StatErrorConfig`, I can only find it in `HistFactorySchema.dtd` file i.e.
```
./HistFactorySchema.dtd:<!ELEMENT Channel (Data*,StatErrorConfig*,Sample+)>
./HistFactorySchema.dtd:<!ELEMENT StatErrorConfig EMPTY>
./HistFactorySchema.dtd:<!ATTLIST StatErrorConfig
```
Okay, seems like this is an obvious difference between the two.

**Reply by jagrundy** [reply]
Just to confirm, there's not a way to do this automatic NP pruning in pyhf like ROOT?

**Reply by alexander-held** [reply]
No, this is currently not technically possible. See https://github.com/scikit-hep/pyhf/issues/662.

**Reply by jagrundy** [reply]
Okay, fair enough. I'm just dropping all stat errors from the pyhf fit to see if this causes a big speed up. If so, then this difference is likely from the pruning, as you say.

**Reply by jagrundy** [reply]
Actually, I've dropped all stat errors from the workspace, but the speed doesn't seem to have improved significantly.

**Reply by lukasheinrich** [reply]
interesting, can you pass us the workspace privately .. we can do some more checks

**Reply by jagrundy** [reply]
Cool, thank you. I can email it over. Who should I address it to?

**Comment by jagrundy** [comment]
Just for some more information about the workspaces, here's the summary of `pyhf inspect`:

| Summary | |
| -------------- | - |
| Channels | 38 |
| Samples | 114 |
| Parameters | 305 |
| Modifiers | 305 |

**Reply by alexander-held** [reply]
How many bins is this in total across the 38 channels?

**Reply by jagrundy** [reply]
Apologies, posted this before I saw your original reply! Anyway, in total there's 564 bins.

**Reply by alexander-held** [reply]
This is a fairly complex setup, I am surprised the ROOT setup manages to converge in 79 seconds. I think to understand this better, we should compare times for a single MLE and then possibly also for more simplified setups. With this many bins and samples, there also may be other strategies that could help, like holding some parameters constant at first and fitting the rest, and then unfreezing all parameters when closer to the minimum.

**Reply by lukasheinrich** [reply]
What backend are you using? Interesting setup for sure, how do you get that high a number of samples? Does this have to do with the lack of expression formulas in pyhf?

**Reply by jagrundy** [reply]
Thank you for the replies. I'm now looking at setting up a comparison of the single MLE fit. To answer the questions in the meantime - we're using jax right now but have tested others in the past and haven't seen a significant difference. Regarding the number of samples, in each channel we have two signal models and background so 38 channels x (2 sig + 1 bkg) = 114 samples.

**Reply by lukasheinrich** [reply]
Do you allocate a different sample per channel? This might be a clue. Pyhf wouldn't optimise that

**Reply by jagrundy** [reply]
Yes, we did. I can look at changing that now. Will let you know the result.

**Reply by jagrundy** [reply]
I've reduced the number of samples to three but there doesn't seem to be any appreciable speed up from an initial test.

**Reply by AlkaidCheng** [reply]
Hi, the default settings for `quickstats` is roughly (in ROOT language):
-minimizer strategy = 1
-minimizer type = Minuit2
-minimizer algorithm = Migrad
-minimizer tolerance = 1.
-constant term optimization + Cache-and-track optimization = True

From my test, the entire limit setting takes 27 seconds to finish for the 4b workspace mentioned.

**Comment by kratsg** [comment]
Hi, there's been no update. @AlkaidCheng is this still an open discussion? Re-open/comment if so.
